easybuild/tools/modules.py

```diff
@@ -543,9 +543,21 @@ def mod_exists_via_show(mod_name):
 
             :param mod_name: module name
             """
-            mod_exists_regex = mod_exists_regex_template % re.escape(mod_name)
             txt = self.show(mod_name)
-            return bool(re.search(mod_exists_regex, txt, re.M))
+            res = False
+            names_to_check = [mod_name]
+            # The module might be an alias where the target can be arbitrary
+            # As a compromise we check for the base name of the module so we find
+            # "Java/whatever-11" when searching for "Java/11" (--> basename="Java")
+            basename = os.path.dirname(mod_name)
+            if basename:
+                names_to_check.append(basename)
```

Comment:
While looking into this, I noticed this change is not required at all to make the enhanced test pass, because of the fallback mechanism with `module_wrapper_exists` in `ModulesTool.show` that was added in #2606 .

So, I'm still puzzled why this change is required at al...

I enhanced the checks in `test_exist` a bit more, and roll back the changes you made to `mod_exists_via_show` here, and it still works fine.

See https://github.com/Flamefire/easybuild-framework/compare/improve_module_exist...boegel:improve_module_exist. 

I can open a PR to your branch with these changes once I get the `test_run_module` I added passing with all different module tools.

Reply:
I propose we roll back this change for now, which is done as a part of https://github.com/Flamefire/easybuild-framework/pull/2 .

We can further extend `test_exist` to cover the use cases that are actually broken (for example module aliases as discussed in #3215), before looking into a fix.

Reply:
> So, I'm still puzzled why this change is required at al...

Because as explained in #3215 the fallback shouldn't be there: It is fuzzy and incomplete. If `module show` returns a module, then the module exists else it does not, no fallback required.

Just looked at why the tests pass and comparing the #3215 and the docu at https://lmod.readthedocs.io/en/latest/093_modulerc.html:

Fallback considers only: 
- `module_version` command
- only module paths and 1 given directory

Fallback hence fails for:
- `module_alias`
- system MODULERC files
- ~/.modulerc.lua or ~/.modulerc

The last one was the problem I found. Users can't change the files in modulepath but they may want to change aliases via a home-folder modulerc.

Counter-proposal: We keep this change and print a fat warning when a module is not found via `module show` but via the fallback and gradually get rid of the fallback (My assumption is the warning will never trigger with this change)

Or do you see any reason to keep and improve a hand-rolled implementation of part of LMod/envModules?

More arguments: The fallback is particularly tricky as you noted in https://github.com/easybuilders/easybuild-framework/pull/2606#issuecomment-426916059 and I'm not sure that this is addressed: If a module wrapper exists but the underlying module does not, what does happen? I can't find such a 2nd check in https://github.com/easybuilders/easybuild-framework/blob/3ff16d7c3fe5378ed92118f372e289f6bd820a98/easybuild/tools/modules.py#L486
Using `module show` only avoids this problem altogether.

To go forward faster: Is there anything wrong with this PR? Any possible failure you can see? Any test that fails with it? If you want a test which passes after but not before I guess using `module_alias` will be enough. Can add that Tuesday.

Reply:
I don't have any strong objections against your approach, other than that it partially solves the same issue as `module_wrapper_exists` does.

I agree that your approach is better, thanks a lot for the clarification.
`module_wrapper_exists` indeed has shortcomings (quite a lot actually, as you point them out), and it's indeed not EasyBuild's task to re-implement the modules tool.

With that in mind, I suggest proceeding as follows:

1) Merge https://github.com/Flamefire/easybuild-framework/pull/2 to update this PR, mostly for the additional tests (which are largely not directly related to the changes made in this PR, but are still useful). I spent quite a bit of time on the test for `run_module` in particular, it would be great if that effort doesn't go to waste...

2) Revert commit a10a6ac that undoes your changes in `mod_exists_via_show` (or do a `git rebase --interactive` to kick it out entirely). I can also kick out that commit from https://github.com/Flamefire/easybuild-framework/pull/2 before you merge it, if you prefer.

3) Add more checks in `test_exist` to cover the use cases that are not dealt with by `module_wrapper_exists`, preferably not only `module_alias` but also the others, if possible. For `~/.modulerc.lua` and `~/.modulerc`, you can probably cover those in the test by temporarily redefining `$HOME`?

4) Remove `module_wrapper_exists` entirely, since it's no longer needed with your enhanced `mod_exists_via_show`. I see no real reason to be careful there, the checks in `test_exist` should be good enough to ensure that nothing gets broken that was covered by `module_wrapper_exists`.

    If you prefer being cautious, I guess adding a `log.deprecated` statement in `module_wrapper_exists` makes sense too, to leave the fallback in place for now but print a clear warning if it's still triggered. If we take this route, the message should clearly mention to report the use case that triggers the warning, so we know we missed something before we go ahead and remove that block of code.

5) Make your enhanced `mod_exists_via_show` a little bit stricter, as follows:
    ```python
    names_to_check.append(basename + '/')
    ```
    The regular expression used to check the output of `module show` is a bit too liberal now in combination with only using the directory part of the module name.

    Using just `Java` in the fallback check would result in checking with `'^\s*\S*/Java.*(\.lua)?:\s*$'` for example, while it should really be `'^\s*\S*/Java/.*(\.lua)?:\s*$'`.
    Maybe we should even use `basename + '/.*'`, and remove the `.*` from the `mod_exists_regex_template` values?
    I'm not sure why it's even there at all, we expect an exact match (except for the `.lua` part), don't we?

6) Rename `basename` to something else, since it causes confusion. You use `os.path.dirname` to define a variable named `basename`, which clashes directly with `os.path.basename` which does the exact opposite (return just the filename). :)

Reply:
Will work on this Tuesday, some comments now to leave time for decisions:

> For ~/.modulerc.lua and ~/.modulerc, you can probably cover those in the test by temporarily redefining $HOME?

I didn't add tests yet for fear of accidentally removing the real file or e.g. missing to undefine it or so... I'll see what I can come up with

> Remove module_wrapper_exists entirely,

I'm all in for that. But what about your usual reasoning someone else might be using it?

> print a clear warning if it's still triggered. If we take this route, the message should clearly mention to report the use case that triggers the warning, so we know we missed something before we go ahead and remove that block of code.

We must not call it within EasyBuild or make the message only trigger if `module_wrapper_exists` returns anything because it is currently called if the previous check does not find anything, which is valid: When the module indeed does not exist. Problem: An alias might exist but the underlying module not. `module show` would return nothing but `module_wrapper_exists` would return something. So I'd never call it from EB and deprecate it like other functions

> Make your enhanced mod_exists_via_show a little bit stricter, as follows:

Why? If `$LMOD_CMD show foo/1.2` returns `bar_3.4.lua` because someone set the alias as such, then this is valid. AFAIK `module show` returns nothing when the module is not found (maybe an error? But even that not on stderr of the module tool as that is up to the wrapper)      
Hence I don't think it needs to be made any stricter. Use case for the loose version: IBM has a custom TensorFlow. Maybe someone makes an alias `TensorFlow_PowerAI` to `TensorFlow/1.13_Anaconda_...` or so to distinguish it from official TF

Do you know of anything where the strict check would be required? Then we should add negative tests too that cover that. But until then I would keep it loose